![New York City schoolbus](schoolbus.jpg)

Photo by [Jannis Lucas](https://unsplash.com/@jannis_lucas) on [Unsplash](https://unsplash.com).
<br>

Every year, American high school students take SATs, which are standardized tests intended to measure literacy, numeracy, and writing skills. There are three sections - reading, math, and writing, each with a **maximum score of 800 points**. These tests are extremely important for students and colleges, as they play a pivotal role in the admissions process.

Analyzing the performance of schools is important for a variety of stakeholders, including policy and education professionals, researchers, government, and even parents considering which school their children should attend. 

You have been provided with a dataset called `schools.csv`, which is previewed below.

You have been tasked with answering three key questions about New York City (NYC) public school SAT performance.

### 1. Load and Inspect data

In [1]:
# Import libraries
import pandas as pd

# Load in the data
schools = pd.read_csv("schools.csv")

# Preview first five rows
schools.head()

school_name    borough building_code  \
0  New Explorations into Science, Technology and ...  Manhattan          M022   
1                               Essex Street Academy  Manhattan          M445   
2                       Lower Manhattan Arts Academy  Manhattan          M445   
3    High School for Dual Language and Asian Studies  Manhattan          M445   
4      Henry Street School for International Studies  Manhattan          M056   

   average_math  average_reading  average_writing  percent_tested  
0           657              601              601             NaN  
1           395              411              387            78.9  
2           418              428              415            65.1  
3           613              453              463            95.9  
4           410              406              381            59.7

In [2]:
# Check the shape
print('Shape of data: ', schools.shape)

# Inspect data structure and types
schools.info()

Shape of data:  (375, 7)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375 entries, 0 to 374
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   school_name      375 non-null    object 
 1   borough          375 non-null    object 
 2   building_code    375 non-null    object 
 3   average_math     375 non-null    int64  
 4   average_reading  375 non-null    int64  
 5   average_writing  375 non-null    int64  
 6   percent_tested   355 non-null    float64
dtypes: float64(1), int64(3), object(3)
memory usage: 20.6+ KB


**Summary**

Based on the output:
- The dataset contains **375 rows and 7 columns**
- There are no missing values and data types are correctly formatted

Since we are provided with a clean data, we will proceed with answering the key questions.


### 2. Finding NYC schools with the best math results

In [3]:
# Define threshold for top math performance (80% of 800)
best_math_result = (80 / 100) * 800

# Filter schools with average math scores above the threshold and sort in descending order
best_math_schools = (
    schools[schools["average_math"] >= best_math_result][["school_name", "average_math"]]
    .sort_values("average_math", ascending=False)
    .head()
)

# Display the top 5 schools with the highest average math scores
best_math_schools

school_name  average_math
88                              Stuyvesant High School           754
170                       Bronx High School of Science           714
93                 Staten Island Technical High School           711
365  Queens High School for the Sciences at York Co...           701
68   High School for Mathematics, Science, and Engi...           683

### 3. Identifying the top ten performing schools

In [4]:
# Calculate total SAT score by summing math, reading, and writing scores
schools["total_SAT"] = schools[["average_math", "average_reading", "average_writing"]].sum(axis=1)

# Sort schools by total SAT score in descending order
top_schools = schools[["school_name", "total_SAT"]].sort_values("total_SAT", ascending=False)

# Select the top 10 schools with the highest total SAT scores
top_10_schools = top_schools.head(10)

# Display the result
top_10_schools

school_name  total_SAT
88                              Stuyvesant High School       2144
170                       Bronx High School of Science       2041
93                 Staten Island Technical High School       2041
174  High School of American Studies at Lehman College       2013
333                        Townsend Harris High School       1981
365  Queens High School for the Sciences at York Co...       1947
5                       Bard High School Early College       1914
280                     Brooklyn Technical High School       1896
45                       Eleanor Roosevelt High School       1889
68   High School for Mathematics, Science, and Engi...       1889

### 4. Locating the NYC borough with the largest standard deeviation in SAT performance

In [5]:
# Group schools by borough and calculate count, mean, and standard deviation of total SAT scores
borough_stats = schools.groupby("borough").agg(
    num_schools=("total_SAT", "count"),
    average_SAT=("total_SAT", "mean"),
    std_SAT=("total_SAT", "std")
).reset_index()

# Round average and standard deviation columns to two decimal places
borough_stats[["average_SAT", "std_SAT"]] = borough_stats[["average_SAT", "std_SAT"]].round(2)

# Identify the borough with the highest standard deviation in SAT scores
largest_std_borough = borough_stats.loc[borough_stats["std_SAT"].idxmax()]

# Convert the result to a one-row DataFrame
largest_std_dev = pd.DataFrame([largest_std_borough])

# Display the borough with the most variation in SAT scores
largest_std_dev

borough  num_schools  average_SAT  std_SAT
2  Manhattan           89      1340.13   230.29